### PR TITLE
Fix a NSDNHAO in extension methods.

### PR DESCRIPTION
**Pull request by retronym**
It is shown up by the overriding and compiling without specialization.

Review by @paulp, @odersky

Discussion: https://groups.google.com/d/topic/scala-internals/8_sxnJhRfyA/discussion

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala [hunk 1]
```diff
@@ -45,7 +45,7 @@ abstract class ExtensionMethods extends Transform with TypingTransformers {
    *  of how overloaded types are ordered between phases and picklings.
    */
   private def extensionNames(imeth: Symbol): Stream[Name] =
-    imeth.owner.info.decl(imeth.name).tpe match {
+    imeth.owner.info.member(imeth.name).tpe match {
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala]: Hmm. Could you explain the difference between `findDecl` and `findMember`? Is it only the fact that `decl` looks up only in the current class, whereas `member` also traverses the inheritance hierarchy?

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala]: That's my (recently gained) understanding. It also respects subsumption, so you don't end up with both the superclass and subclass methods. That's was I thought was happening here, but that's not right at all, as `decl`  can only be giving us members from `StringOps` directly.

Looking at it further: we end up with this type:

```
AnyVal
        with T {
  val repr0: String
  private[this] val repr0: String
  def <init>(repr0: String): StringOps
  def x: Unit
  override def hashCode(): Int
  override def equals(x$1: Any): Boolean
  def x(): Object
}
```

`def x: Unit` is excluded by `member`, but included by `decl`. But the real question is, where does it come from? (That probably has an easy answer, but I don't know it.)

Comment (paulp) [inline src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala]: If there's another intentional difference I'm not aware of it.

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala]: Please ignore my babbling. `def x: Unit` is the just method I defined.

**Comment by retronym** [comment]
The material difference here between `decl` and `member` is that the latter actively excludes bridge methods, as added by erasure when an overriding method refines a return type.

``` scala
    /** The defined or declared members with name `name` in this type;
     *  an OverloadedSymbol if several exist, NoSymbol if none exist.
     *  Alternatives of overloaded symbol appear in the order they are declared.
     */
    def decl(name: Name): Symbol = findDecl(name, excludedFlags = 0)

    /** The member with given name,
     *  an OverloadedSymbol if several exist, NoSymbol if none exist */
    def member(name: Name): Symbol =
      memberBasedOnName(name, excludedFlags = BridgeFlags)
```
- Why is the bridge method visible at all, given that we are running underneath `atPhase(currentRun.refChecksPhase)`?

**Comment by retronym** [comment]
### With Specialization

#### Type History

```
TypeHistory(posterasure:1,Object
        with T {
  val repr0: String
  private[this] val repr0: String
  def <init>(repr0: String): StringOps
  def x: Unit
  override def hashCode(): Int
  override def equals(x$1: Any): Boolean
  def x(): Object
},TypeHistory(explicitouter:1,AnyVal
        with T {
  val repr0: String
  private[this] val repr0: String
  def <init>(repr0: String): StringOps
  def x: Unit
  override def hashCode(): Int
  override def equals(x$1: Any): Boolean
  def x(): Object
},TypeHistory(namer:1,AnyVal
        with T {
  val repr0: String
  private[this] val repr0: String
  def <init>(repr0: String): StringOps
  def x: Unit
  override def hashCode(): Int
  override def equals(x$1: Any): Boolean
},null)))
```

#### atPhase(refchecks)

```
AnyVal
        with T {
  val repr0: String
  private[this] val repr0: String
  def <init>(repr0: String): StringOps
  def x: Unit
  override def hashCode(): Int
  override def equals(x$1: Any): Boolean
}
```

### Without specialization

#### Type History

```
TypeHistory(posterasure:1,Object
        with T {
  val repr0: String
  private[this] val repr0: String
  def <init>(repr0: String): StringOps
  def x: Unit
  override def hashCode(): Int
  override def equals(x$1: Any): Boolean
  def x(): Object
},TypeHistory(namer:1,AnyVal
        with T {
  val repr0: String
  private[this] val repr0: String
  def <init>(repr0: String): StringOps
  def x: Unit
  override def hashCode(): Int
  override def equals(x$1: Any): Boolean
  def x(): Object
},null))
```

#### atPhase(refchecks)

```
AnyVal
        with T {
  val repr0: String
  private[this] val repr0: String
  def <init>(repr0: String): StringOps
  def x: Unit
  override def hashCode(): Int
  override def equals(x$1: Any): Boolean
  def x(): Object
}
```

**Comment by paulp** [comment]
My guess is we're looking at the difference between "setInfo" and "updateInfo".

**Comment by odersky** [comment]
It's a bit tricky, because extension methods are only created for the definitions in the value class itself, not for any inherited traits. We can't generate for inherited traits because in general we do not have their code in hand. That's why we look for `decl`, not for `member`. I have to look at the whole logic to see whether the fix is correct.

**Comment by odersky** [comment]
I think the discussion wrt bridges is spot on. We still want to take `decl` in `extensionNames`, but exclude bridges.

**Comment by odersky** [comment]
I figured out why it only shows up without specialization. Bridge generation is done at phase "erasure", but new scopes are only generated for the phase after that. So bridges are visible in earlier phases. However, specialization does create new scopes (I believe that's the explicitOuter scope we see), so bridges are only visible from then on, but not before.

An alternative would be to create new scopes already at erasure phase, so bridges would not be visible before. However, I believe we would still see bridges from inherited classes, because they get read in from Java class files with the rest. That's why the current scheme is to leave bridges in scope but filter them out again in findMember. (and also, copying scopes is not free, of course),.

I remember there were several iterations wrt bridges early in Scala's life, and it was difficult to get them right. So we could give it another try, but it would risk upsetting lots of things.

So, clearly, the correct fix for the current problem is to leave it at `decl` but filter out bridges.

**Pull request closed without merging**